src/main/java/com/conveyal/r5/analyst/Grid.java

```diff
@@ -126,96 +126,154 @@ public Grid (int zoom, int width, int height, int north, int west) {
      * the weight at a pixel is the proportion of the input geometry that falls within that pixel.
      */
-    public TObjectDoubleMap<int[]> getPixelWeights (Geometry geometry) {
-        return getPixelWeights(geometry, false);
+    public ArrayList<PixelWeight> getPixelWeights (Geometry geometry) {
+        return getPixelWeights(geometry, false, null);
     }
 
     /**
-     * Get the proportions of an input polygon feature that overlap each grid cell, in the format [x, y] => weight.
-     * This weight object can then be fed into the incrementFromPixelWeights function to actually burn a polygon into the
+     * Get the proportions of an input polygon feature that overlap each grid cell, for use in List(x, y, weight)
+     * These weight lists can then be fed into the incrementFromPixelWeights function to actually burn a polygon into the
      * grid.
      *
      * If relativeToPixels is true, the weights are the proportion of the pixel that is covered. Otherwise they are the
-     * portion of this polygon which is within the given grid cell. If using incrementPixelWeights, this should be set to
+     * portion of this polygon which is within the given pixel. If using incrementPixelWeights, this should be set to
      * false.
      *
-     * This returns a map from int arrays containing the coordinates to the weight. I believe int arrays in hashmaps are
-     * hashed based on memory location, meaning lookups will fail. However, we only use this in a streaming fashion, looping
-     * over all values, which is harmless.
+     * This used to return a map from int arrays containing the coordinates to the weight.
      *
-     * TODO create a class with x, y, and value attributes and just return a list.
      */
-    public TObjectDoubleMap<int[]> getPixelWeights (Geometry geometry, boolean relativeToPixels) {
+    private ArrayList<PixelWeight> getPixelWeights (Geometry geometry, boolean relativeToPixels, PixelWeightCallback pixelWeightCallback) {
         // No need to convert to a local coordinate system
         // Both the supplied polygon and the web mercator pixel geometries are left in WGS84 geographic coordinates.
         // Both are distorted equally along the X axis at a given latitude so the proportion of the geometry within
         // each pixel is accurate, even though the surface area in WGS84 coordinates is not a usable value.
 
-        TObjectDoubleMap<int[]> weights = new TObjectDoubleHashMap<>();
+        ArrayList<PixelWeight> weights = new ArrayList<>();
 
         double area = geometry.getArea();
         if (area < 1e-12) {
             throw new IllegalArgumentException("Geometry is too small");
         }
 
+        // PreparedGeometry is often faster for small numbers of vertices;
+        // see https://github.com/chrisbennight/intersection-test
+
+        PreparedGeometryFactory pgFact = new PreparedGeometryFactory();
```

Comment:
I wonder if we're really intended to construct a new factory for each single prepared geometry we create. It might be sufficient to have one factory instance in a static field on this class, assuming that the factory is threadsafe.